src/basics.md

```diff
@@ -900,6 +901,62 @@ fn run() -> Result<()> {
 # quick_main!(run);
 ```
 
+[ex-pbkdf2]: #ex-pbkdf2
+<a name="ex-pbkdf2"></a>
+## Salt and hash a password with PBKDF2
+
+[![ring-badge]][ring] [![cat-cryptography-badge]][cat-cryptography]
+
+Uses [`ring::pbkdf2`] to hash a salted password using the PBKDF2 key derivation
+function [`ring::pbkdf2::derive`] and then verifies the hash is correct with
+[`ring::pbkdf2::verify`].
+
+```rust
+# #[macro_use]
+# extern crate error_chain;
+extern crate data_encoding;
+extern crate ring;
+#
+# error_chain! {
+#   foreign_links {
+#     Ring(ring::error::Unspecified);
+#   }
+# }
+
+use data_encoding::HEXUPPER;
+use ring::{digest, pbkdf2, rand};
+use ring::rand::SecureRandom;
+
+fn run() -> Result<()> {
+  let rng = rand::SystemRandom::new();
+  static DIGEST_ALG: &'static digest::Algorithm = &digest::SHA512;
+  const CREDENTIAL_LEN: usize = digest::SHA512_OUTPUT_LEN;
+  const N_ITER: u32 = 100_000;
+
+  let mut salt = [0u8; CREDENTIAL_LEN];
+  let mut pbkdf2_hash = [0u8; CREDENTIAL_LEN];
+  rng.fill(&mut salt)?;
+
+  // Hash password
+  let password = "password";
+  pbkdf2::derive(DIGEST_ALG, N_ITER, &salt, password.as_bytes(),
+    &mut pbkdf2_hash);
+  println!("PBKDF2 hash: {}", HEXUPPER.encode(&pbkdf2_hash));
+
+  // Verify hash
+  let hash_verification = pbkdf2::verify(DIGEST_ALG, N_ITER, &salt,
+    password.as_bytes(), &pbkdf2_hash);
+  match hash_verification {
+    Ok(_) => println!("Successfully verified password"),
```

Comment:
How about we provide two separate calls to verify one with correct password and second one with bogus one each succeeded by call to `assert!(hash_verification.is_ok());` instead of the `match` and `println!`?

I'm not sure what would be nicer here. So I'll let you decide.

Reply:
I added the asserts, however on the failing assert I do `assert!(!should_fail.is_ok());` so we don't have to worry about a failing assert. I have encountered differing opinions on this approach, so I am happy to remove the negation if you prefer!